### PR TITLE
Consecutive Absences and Last Service Date fields population

### PR DESCRIPTION
**Pull request by bethbrains**
# Critical Changes

# Changes
- Contact and Program Engagement now have a field called "Last Service Date" automatically populated (when the associated Feature Gates are active). This will be the Delivery Date of the most recent Service Delivery that has a Quantity of >0. 
- Contact and Program Engagement now have a field called "Consecutive Absences" automatically populated (when the associated Feature Gates are active). This will count back from today and count up how many Absent Service Delivery records there are before the first Present record.

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [x] Any net new LWC work has Sa11y tests & 50% or above lines JEST test coverage  
- [x] CRUD/FLS is enforced in Apex
- [x] Permission sets are updated to account for CRUD|FLS|Tab|Classes
- [x] Field sets are updated to account for new fields
- [x] UX approval or UX not necessary
- [x] Link the pull request and work item by PR comment and Chatter post respectively, e.g. GUS: W-0000000: Work Name (https://github.com/SalesforceFoundation/PMM/pull/303)
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [x] PR contains draft release notes
- [ ] QE story level testing completed

### CODE REVIEW

**Comment by bethbrains** [comment]
[W-8827100](https://gus.my.salesforce.com/a07B00000095N2LIAU)

**Review by jjbennett — COMMENTED** [review]
This is looking really great, one early exit suggestion and a concern with the creating default target object and nulling out the consecutive absence field. Maybe we could right a quick test for that scenario? Thanks!

File: force-app/main/default/classes/AttendanceService.cls [hunk 1]
```diff
@@ -258,20 +366,29 @@ public with sharing class AttendanceService {
...
-            SObject targetObject = targetObjectById.get(targetId);
+            SObject targetObject = rollupTargetObjectById.get(targetId);
             if (targetObject == null) {
                 targetObject = targetId.getSObjectType().newSObject(targetId);
-                targetObjectById.put(targetId, targetObject);
+                rollupTargetObjectById.put(targetId, targetObject);
                 targetObject.put(attendedField, null);
                 targetObject.put(absentField, null);
+                if (consecutiveAbsenceField != null) {
+                    targetObject.put(consecutiveAbsenceField, null);
```
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/classes/AttendanceService.cls:390]: OMG my brain is fried, thanks for clearing that up!

File: force-app/main/default/classes/AttendanceService.cls [hunk 2]
```diff
@@ -79,46 +99,134 @@ public with sharing class AttendanceService {
...
+                    lookupFieldOnDelivery,
+                    idsByLookupFieldToTarget.get(lookupFieldOnDelivery),
+                    parentObject,
+                    relationshipName
+                )
+            ) {
+                SObject targetObject = rollupTargetObjectById.get(parentRecord.Id);
+                if (targetObject == null) {
+                    targetObject = parentRecord;
+                    rollupTargetObjectById.put(targetObject.Id, targetObject);
```
[the remark this answers is not in the document]
Comment (jjbennett) [reply force-app/main/default/classes/AttendanceService.cls:214]: Part of my earlier confusion lol.

**Pull request merged**